### PR TITLE
Feature: support snell v2 (without connection reuse)

### PR DESCRIPTION
**Pull request by kamingchan**
#920 

base on https://github.com/icpz/snell-server-reversed and https://github.com/icpz/open-snell

### CODE REVIEW

File: component/snell/cipher.go [hunk 1]
```diff
@@ -1,21 +1,59 @@
...
 
 import (
+	"crypto/aes"
 	"crypto/cipher"
 
 	"golang.org/x/crypto/argon2"
+	"golang.org/x/crypto/chacha20poly1305"
 )
 
+type Cipher interface {
```
Comment (Dreamacro) [inline component/snell/cipher.go:11]: aka `github.com/Dreamacro/go-shadowsocks2/shadowaead` shadowaead.Cipher

File: component/snell/cipher.go [hunk 2]
```diff
@@ -1,21 +1,59 @@
...
-	return sc.makeAEAD(argon2.IDKey(sc.psk, salt, 3, 8, 1, uint32(sc.KeySize())))
+	return sc.makeAEAD(snellKDF(sc.psk, salt, sc.KeySize()))
 }
 func (sc *snellCipher) Decrypter(salt []byte) (cipher.AEAD, error) {
-	return sc.makeAEAD(argon2.IDKey(sc.psk, salt, 3, 8, 1, uint32(sc.KeySize())))
+	return sc.makeAEAD(snellKDF(sc.psk, salt, sc.KeySize()))
+}
+
+func snellKDF(psk, salt []byte, keySize int) []byte {
+	return argon2.IDKey(psk, salt, 3, 8, 1, 32)[:keySize]
```
Comment (Dreamacro) [inline component/snell/cipher.go:34]: add a comment `// snell use a special kdf function` in the previous line

File: component/snell/snell.go [hunk 3]
```diff
@@ -70,12 +77,16 @@ func (s *Snell) Read(b []byte) (int, error) {
...
 }
 
-func WriteHeader(conn net.Conn, host string, port uint) error {
+func WriteHeader(conn net.Conn, host string, port uint, version int) error {
 	buf := bufferPool.Get().(*bytes.Buffer)
 	buf.Reset()
 	defer bufferPool.Put(buf)
 	buf.WriteByte(Version)
-	buf.WriteByte(CommandConnect)
+	if version == Version2 {
```
Comment (Dreamacro) [inline component/snell/snell.go:85]: snell v2 always write version1 (I don't know why, but it's)

**Review by Dreamacro — CHANGES_REQUESTED** [review]

File: component/snell/snell.go (hunk 3 above)
Comment (kamingchan) [reply component/snell/snell.go]: Fine. Both version1 and version2 work on official snell server.

**Comment by Dreamacro** [comment]
Add pool for version 2, thanks for your contribution

**Comment by itspig** [comment]
请问配置文件是snell节点加
version: 2 
就可以变成 v2 版本了吗？

**Pull request merged**